java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -573,7 +573,10 @@ private void refineType(AbstractTypedTree expression, JavaType expressionType, J
           .findFirst()
           .ifPresent(t -> {
             if(refinedReturnType instanceof DeferredType) {
-              setInferedType(functionType.typeSubstitution.substitutedType(t), (DeferredType) refinedReturnType);
+              DeferredType deferredType = (DeferredType) refinedReturnType;
+              if(deferredType.tree() != null) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png 'Severity: CRITICAL') Refactor this code to not nest more than 3 if/for/while/switch/try statements. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS134)